keps/sig-network/3766-referencegrant/README.md

```diff
@@ -0,0 +1,688 @@
+kind: ReferenceGrant
+metadata:
+  name: allow-ns1-gateways-to-ref-secrets
+  namespace: gateway-api-example-ns2
+spec:
+  from:
+  - group: gateway.networking.k8s.io
+    resource: gateways
+    namespace: gateway-api-example-ns1
+  to:
+  - resource: secrets
+```
+
+For Gateway TLS references, if this ReferenceGrant is deleted (revoking, 
+the grant), then the Listener will become invalid, and the configuration
+will be removed as soon as possible (eventual consistency permitting).
+
+#### Gateway API HTTPRoute Referencing Service
+
+In this example, a HTTPRoute in the `baz` namespace is directing traffic
+to a Service backend in the `quux` namespace.
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: quuxapp
+  namespace: baz
+spec:
+  parentRefs:
+  - name: example-gateway
+    sectionName: https
+  hostnames:
+  - quux.example.com
+  rules:
+  - matches:
+    - path:
+        type: PathPrefix
+        value: /
+    backendRefs:
+  # BackendRefs are Services by default.
+    - name: quuxapp
+    namespace: quux
+      port: 80
+---
+apiVersion: authorization.k8s.io/v1alpha1
+kind: ReferenceGrant
+metadata:
+  name: allow-baz-httproutes
+  namespace: quux
+spec:
+  from:
+  - group: gateway.networking.k8s.io
+    resource: httproutes
+    namespace: baz
+  to:
+  - resource: services
+```
+
+For HTTPRoute objects referencing a backend in another namespace, if the
+ReferenceGrant is deleted, the backend will become invalid (since the target
+can't be found). If there was more than one backend, then the valid parts of the
+HTTPRoute's config would persist in the data plane.
+
+But in this case, the cross-namespace reference is the _only_ backend, so the
+removal of the ReferenceGrant will also result in the removal of the HTTPRoute's
+config from the data plane.
+#### PersistentVolumeClaim using cross namespace data source
+
+This example is taken from https://kubernetes.io/blog/2023/01/02/cross-namespace-data-sources-alpha/
+and updated to use the proposed new spec.
+
+It allows the PersistentVolumeClaim in the `dev` namespace to use a volume
+snapshot from the `prod` namespace as its data source.
+
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: example-pvc
+  namespace: dev
+spec:
+  storageClassName: example
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+  dataSourceRef:
+    apiGroup: snapshot.storage.k8s.io
+    kind: VolumeSnapshot
+    name: new-snapshot-demo
+    namespace: prod
+  volumeMode: Filesystem
+---
+apiVersion: authorization.k8s.io/v1alpha1
+kind: ReferenceGrant
+metadata:
+  name: allow-prod-pvc
+  namespace: prod
+spec:
+  from:
+  - resource: persistentvolumeclaims
+    namespace: dev
+  to:
+  - group: snapshot.storage.k8s.io
+    resource: volumesnapshots
+    name: new-snapshot-demo
+```
+
+At time of writing, I'm not sure about the behavior here when the ReferenceGrant
+is deleted.
```

Comment:
@msau42, I didn't know where to go to find this out, if you could comment here, I'll update with any details you can provide.

Reply:
@ttakahashi21 can you help provide more context here?

Reply:
https://github.com/kubernetes/enhancements/pull/3767/files#r1084757612
It works as described in the comment above.
It is only used when creating a PVC. So, revoking access would not impact the existing volume.
Only new volumes would be rejected.

Reply:
I think the comment got lost after uploading a new set of changes. Copying it over here:

For the volume provisioning from datasource scenario, once the volume is provisioned, it no longer actively uses the reference, so revoking access would not impact the existing volume. Only new volumes would be rejected.

Reply:
Thanks, clarified in the KEP, PTAL